pkg/seven/state/builder.go

```diff
@@ -84,161 +86,198 @@ func paddingUpstream(upstream *v1.Upstream, currentUpstream *v1.Upstream) {
-						// 2. sync memDB
-						us := []*v1.Upstream{u}
-						if !needToUpdate {
-							currentUpstream.Nodes = u.Nodes
-							us = []*v1.Upstream{currentUpstream}
-						}
-						upstreamDB := &db.UpstreamDB{Upstreams: us}
-						if err := upstreamDB.UpdateUpstreams(); err != nil {
-							glog.Errorf("solver upstream failed, update upstream to local db failed, err: %s", err.Error())
-							return
-						}
+				}
+				if needToUpdate {
+					// 1.sync memDB
+					upstreamDB := &db.UpstreamDB{Upstreams: []*v1.Upstream{u}}
+					if err := upstreamDB.UpdateUpstreams(); err != nil {
+						glog.Errorf("solver upstream failed, update upstream to local db failed, err: %s", err.Error())
+						errNotify = err
+						return
 					}
-				} else {
-					op = Create
-					// 1.sync apisix and get response
-					if upstreamResponse, err := apisix.AddUpstream(u); err != nil {
+					// 2.sync apisix
+					if err = apisix.UpdateUpstream(u); err != nil {
 						glog.Errorf("solver upstream failed, update upstream to etcd failed, err: %+v", err)
+						errNotify = err
 						return
-					} else {
-						tmp := strings.Split(*upstreamResponse.Upstream.Key, "/")
-						*u.ID = tmp[len(tmp)-1]
 					}
-					// 2.sync memDB
-					//apisix.InsertUpstreams([]*v1.Upstream{u})
-					upstreamDB := &db.UpstreamDB{Upstreams: []*v1.Upstream{u}}
-					upstreamDB.InsertUpstreams()
 				}
+				// if fromKind == WatchFromKind
+				if u.FromKind != nil && *u.FromKind == WatchFromKind {
+					// 1.update nodes
+					if err = apisix.PatchNodes(u, u.Nodes); err != nil {
+						glog.Errorf("solver upstream failed, patch node info to etcd failed, err: %+v", err)
+						errNotify = err
+						return
+					}
+					// 2. sync memDB
+					us := []*v1.Upstream{u}
+					if !needToUpdate {
+						currentUpstream.Nodes = u.Nodes
+						us = []*v1.Upstream{currentUpstream}
+					}
+					upstreamDB := &db.UpstreamDB{Upstreams: us}
+					if err := upstreamDB.UpdateUpstreams(); err != nil {
+						glog.Errorf("solver upstream failed, update upstream to local db failed, err: %s", err.Error())
+						errNotify = err
+						return
+					}
+				}
+			} else {
+				op = Create
+				// 1.sync apisix and get response
+				if upstreamResponse, err := apisix.AddUpstream(u); err != nil {
+					glog.Errorf("solver upstream failed, update upstream to etcd failed, err: %+v", err)
+					errNotify = err
+					return
+				} else {
+					tmp := strings.Split(*upstreamResponse.Upstream.Key, "/")
+					*u.ID = tmp[len(tmp)-1]
+				}
+				// 2.sync memDB
+				//apisix.InsertUpstreams([]*v1.Upstream{u})
+				upstreamDB := &db.UpstreamDB{Upstreams: []*v1.Upstream{u}}
+				upstreamDB.InsertUpstreams()
 			}
 		}
-		glog.V(2).Infof("solver upstream %s:%s", op, *u.Name)
-		// anyway, broadcast to service
-		serviceWorkers := swg[*u.Name]
-		for _, sw := range serviceWorkers {
-			event := &Event{Kind: UpstreamKind, Op: op, Obj: u}
-			sw.Event <- *event
-		}
+	}
+	glog.V(2).Infof("solver upstream %s:%s", op, *u.Name)
```

Comment:
Should change the log from glog to `pkg/log`.

Reply:
Done